libsolidity/ast/ASTAnnotations.h

```diff
@@ -177,7 +183,7 @@ struct CallableDeclarationAnnotation: DeclarationAnnotation
 	std::set<CallableDeclaration const*> baseFunctions;
 };
 
-struct FunctionDefinitionAnnotation: CallableDeclarationAnnotation, StructurallyDocumentedAnnotation
+struct FunctionDefinitionAnnotation: CallableDeclarationAnnotation, StructurallyDocumentedAnnotation, SharedContractDependenciesAnnotation
```

Comment:
I don't think this is needed. The call graph should take care about it.

Reply:
We'll need to make the graph keep track of inherited contracts.

Reply:
No, I think we need to remove inheritance from the dependencies.

Reply:
Because I think it creates a dependency. Here is how I understand it: if `C is D` then you cannot really compile `D` without being able to finish the compilation of `C` first, even though `D` does not need to access the bytecode of `C` (it has its own "copy"). If we do not detect this, we won't notice that this is a dependency cycle:
- `C` does `new A`
- `A` does `new D`
- `D is C`

Reply:
I don't think this is a problem. All the dependencies that are created are recorded in the call graph. The `new A` inside `C` will be part of `D`'s call graph if it is reachable. If `D` overrides the function in `C` that contains `new A`, then there will be no creation cycle.

Reply:
Oh, you're right. I did not consider that all the stuff from `C` is inherited so it actually becomes a direct cycle between `A` and `D`.

Reply:
Wait, are we talking about `SharedContractDependencyAnnotation` as in, the `contractDependencies` field for `FunctionDefinitionAnnotation` (and not for the `ContractAnnotation`) or is the idea to completely remove `contractDependencies` from all annotations?

Reply:
I don't think we need a field in the function definition annotation. contractDependencies should be computed only from the call graph, though.

Reply:
At the moment, if we don't fill that field from the `TypeChecker`, test cases like this will fail:

```
  Contract:
    contract C { function foo() internal { new D(); } }
    contract D { function foo() internal { new C(); } }

  Expected result:
    TypeError 7813: (91-96): Circular reference found.
  Obtained result:
    Success

```

I guess because neither function is currently reachable.